### ISSUE TITLE
Keyboard shortcuts for copy / paste into console

### ISSUE THREAD

**Issue by MattiasFestin** [Microsoft/WSL#235]
It would be nice to have keyboard shortcut for pasting into console instead of right clicking (Or window menu -> edit -> paste). 

I tend to use my keyboard as much as possible. So it would be nice to have a shortcut like git bash:
- copy = ctrl + insert
- paste = shift + insert

**Comment by benhillis** [comment]
Agreed, this feedback is more general Windows cmd.exe than bash-specific.  Here's a workaround to make copying and pasting easier for you:

```
Launch cmd.exe
Right click title bar, click Properties
Options Tab
Enable QuickEdit mode
```

Now you can highlight text with your mouse cursor, right click to copy, right click again to paste.

**Comment by Birssan** [comment]
`Agreed, this feedback is more general Windows cmd.exe than bash-specific`

That's not entirely true, by default the new cmd.exe accepts text pasted with CTRL + V, it's only once you're in bash mode you can't paste anything with CTRL + V (not even CTRL + Shift + V like in Ubuntu terminal).

![bash-copy](https://cloud.githubusercontent.com/assets/5689193/14701301/fe812a6c-07a4-11e6-9607-a009a45cdefa.PNG)

That's just an option in the properties (`Enable shortcuts with Ctrl` or something like that, my interface is not in English :) )

**Comment by mgleadow** [comment]
I think using the Ubuntu Style Ctrl-Shift-C would make sense, especially as Ctrl-C is a common interrupt

I'd also like to see things like emacs mode honoured where Ctrl-A / Ctrl-K work as expected

**Comment by mobluse** [comment]
AFAIK Ctrl+A and Ctrl+K already works as expected.

**Comment by mgleadow** [comment]
@mobluse So they do.  They didn't work for me the other day though, odd.

**Comment by MattiasFestin** [comment]
> I think using the Ubuntu Style Ctrl-Shift-C would make sense, especially as Ctrl-C is a common interrupt

My toughs where the same with the Ctrl-C (and also Ctrl-V, not that common) interrupt signals. 

That is why I suggested Ctrl + insert and Shift + insert. They work in windows now, so it is nothing new.
And they are compatible with the Linux keyboard interrupts.

**Comment by pawelmhm** [comment]
I know this is issue about copy-paste into console but what's the status of copy-paste from console? I'd like to copy paste output of my terminal interactions into browser. Highlighting text and clicking right mouse button doesn't work (there is no menu shown). Is this some peculiar bug that I encounter with my setup or is it something not supported yet?

**Comment by mirao** [comment]
I vote for Ubuntu terminal shortcuts `Ctrl + Shift + C/X/V` as a primary ones and maybe an alternative `Ctrl/Shift + Insert`. But that's only my preference because I use bash on Linux in most cases.

**Comment by XaviTorello** [comment]
Why not to provide a way to configure the terminal shortcuts for this app? I'm not a win user, i'm just reviewing it

If not, @mirao proposal +1. I'm a *nix user in most cases too, and would be great maintain the default terminal shortcuts for the major distros.

**Comment by GrimTheReaper** [comment]
I would like to have CTRL + Shift + C/X/V, although if possible my vote would be for it to be configurable.

**Comment by bitcrazed** [comment]
This is a current limitation in the Windows Console.

Console currently supports two input modes: 'processed' and 'raw'. In 'processed' mode, Windows Console pre-processes all keystrokes, handling several well-known chords consistently across cmd, PowerShell and elsewhere.

To better handle *NIX behavior, however Bash requests raw, un-pre-processed keyboard input. We do have a task on our backlog to add a "mildly-pre-processed" mode wherein Console will translate copy and paste key chords into the VT sequences expected by *NIX tools, shells, etc.

**Comment by ranyyy** [comment]
You can run `REG.EXE add HKCU\Console /v QuickEdit /t REG_DWORD /d 1 /f` as administrator in the Windows Command Prompt to enable QuickEdit.

This would let you:
- copy text by highlighting text (left clicking) and then right clicking.
- paste text by right clicking

To disable QuickEdit run `REG.EXE add HKCU\Console /v QuickEdit /t REG_DWORD /d 0 /f`.

**Comment by Pixinn** [comment]
I just installed "_Bash on Windows_" and I can copy / paste by doing
* CTRL + SHIFT + C
* Right Click

**Comment by konnichimade** [comment]
Why not Ctrl-Ins/Shift-Ins? Windows 10 CMD honors those and those are also common for SSH.

**Comment by brnelsons** [comment]
ctrl+shift+c
ctrl+shift+v
is what is preferred

**Comment by konnichimade** [comment]
Here's another vote for Ctrl+Ins/Shift+Ins. This seems to be the standard for TTY apps, which is how I usually interact with Linux when in Windows.

**Comment by jskopek** [comment]
+1 to Ctrl-V or some variant. Right clicking to paste works, but taking my hands off the keyboard is far from ideal.

**Comment by farahfa** [comment]
I agree that this should be configurable.. I, for one, would like to have either CTRL+SHIFT+C/X/V or ALT+C/X/V (to simulate how things are done in Mac, the Command + C/X/V)... Also, it would be preferable to also configure the right click to paste as I'd expected to give me a menu (to copy/cut/paste.. etc).. This is especially noticeable when using VIM.. :/ )

**Comment by Charnarongthumwan** [comment]
Good
Thank you.

**Comment by blakehawkins** [comment]
Suggest seeing these uservoice requests in addition to upvoting here:

https://wpdev.uservoice.com/forums/266908-command-prompt-console-bash-on-ubuntu-on-windo/suggestions/13341399-bash-copy-paste

https://wpdev.uservoice.com/forums/266908-command-prompt-console-bash-on-ubuntu-on-windo/suggestions/13373574-bring-ctrl-shift-c-v-to-copy-and-paste

https://wpdev.uservoice.com/forums/266908-command-prompt-console-bash-on-ubuntu-on-windo/suggestions/13389489-better-hotkey-shortcut-integration

**Comment by brennanmh** [comment]
Another vote for gnome-terminal style cut and paste.

The main problem is that it is using the Windows Console app, which, at least from a linux guy, sucks, and always has.

If you really want to cater to linux folks, you need a *much* better terminal emulator.

**Comment by TestSmirk** [comment]
mouse right click is not better than ctrl+shift+c\v.hope ASAP to support this function.

**Comment by ranyyy** [comment]
Well, you can try this program I found. It works great! https://github.com/mintty/wsltty

**Comment by n3rd4i** [comment]
There is a workaround to the WSL Bash, paste issue:

One can simply to the following steps:
1. open cmd.exe
2. bash -c "pasted command here" ( this works because you are in cmd.exe )

The bash will exit after the command, I don't know how, yet, to keep it running after the "variable assign" from the cmd console. e.g. 'bash -ic "command and remain'

**Comment by florianmonfort** [comment]
So @bitcrazed this is in you guys backlog right? Any expectation on release date/time? Coming to a next major version or just simple Windows Update?

I actually "assumed" this would be solved with the Creators Update, and to my surprise it's not actually there. So wondering if we're expecting a time frame in years or in weeks lol

**Comment by kavdev** [comment]
Agreed; @bitcrazed some sort of release timeframe would be awesome.

**Comment by plagov** [comment]
I was unpleasantly surprised that Bash on Windows doesn't support `Shirt-Insert` shortcut. This missing feature is the one that preventing me from switching from the Cygwin. Hope it will really be implemented soon.

**Comment by mobluse** [comment]
@plagov You could use WSLTTY to get Shift+Insert now: https://github.com/mintty/wsltty

**Comment by plagov** [comment]
@mobluse just installing `wsltty-0.7.6-install.exe` and that's it?
Installed it, but still, `Shift-Insert` doesn't work. Or some extra configuration is required?

**Comment by mobluse** [comment]
@plagov I think I didn't configure it, but it can be configured in Options/Keys (Alt+Space O Shift+Tab K).

**Comment by stenrose** [comment]
Another vote for Ctrl-Shift-C and Ctrl-Shift-V copy/paste style.

**Comment by loxal** [comment]
This is a really critical issue. Can you prioritize please?

I already installed ConEmu to workaround this issue but ConEmu's bloated pseudo-feature set is just broken-by-design in so many other ways...

**Comment by sgoranson** [comment]
For now, I have been working around the issue with the following AutoHotkey script:

```
#IfWinActive, ahk_class ConsoleWindowClass
^+v::Rbutton
#ifWinActive
```

**Comment by loxal** [comment]
@sgoranson I've installed https://autohotkey.com/download/ compiled `your-script.ahk` and started the created `your-script.exe` file but still Bash keeps receiving `^V`.

**Comment by loxal** [comment]
What actually worked for me, was:

```
#IfWinActive ahk_class ConsoleWindowClass
^V::
SendInput {Raw}%clipboard%
return
#IfWinActive
```

**Comment by sgoranson** [comment]
@loxal Good call. My version will do annoying things if the cursor is not over cmd.exe.

If anyone wants to use loxal's script with ConEmu, just change ConsoleWindowClass to VirtualConsoleClass

**Comment by nimitbhargava** [comment]
Please do it! I can't click right always to paste. This is not how it should be. Using *Bash on Ubuntu on Windows*

Isn't this something that should come out of box and without requesting. 
Also I find 
![image](https://user-images.githubusercontent.com/8358694/28753796-97731fb6-7557-11e7-8ffa-949576d9656f.png)
but **heck** they don't work!

**Comment by const-g** [comment]
@loxal for me the script didn't work. Indeed, the pasted data replaced ` ` ` characters or dashes.
I had to change it by ```#IfWinActive ahk_class ConsoleWindowClass
^V::
    MouseClick, right
Return
#IfWinActive```

**Comment by xdaxdb** [comment]
The community can debate ctrl+c/v variations as much as needed. But I feel that the faithful paste method of **Shift+Insert** should function as it does in every other terminal application.

Using Ctrl+A as an alternative to Ctrl+C is _awful_ because it conflicts with the Bash Ctrl+A/E beginning/end of line cursor placement shortcuts which are huge timesavers when repeating/modifying long commands. **Ctrl+A would break that feature.**

**Comment by spirit986** [comment]
I have the same problem ... CTRL+Insert and SHIFT+Insert will not work and also CTRL+SHIFT+c/v does not work. I need to use the mouse to paste and this is incredibly annoying. Only one way to copy paste. At the same time most of the other functions work.

**Comment by stianl** [comment]
wsltty is a great suggestion for enabling keyboard shortcuts for copy/paste for wsl (and it uses the sensible CTRL/SHIFT+INSERT shortcuts). It will also enable ALT-SPACE to select the window menu (and if you use it for nano or other things that depends on it you can still turn it off from a simple option panel). 

However, it does not provide the nice [text selection features](https://technet.microsoft.com/library/mt427362.aspx) (SHIFT-ARROW keys to select text) that the console has (and that it "loses" under wsl) 😢

**Comment by radistao** [comment]
Vote up 
+1
Please

**Comment by ComputerDoktor** [comment]
Vote up
+1
Please

**Comment by bitcrazed** [comment]
We're very (painfully!) aware of this ask. 

Sounds like it'd be the simplest feature to implement, right? I mean, how hard could adding a handler for CTRL+SHIFT+V be? Turns out there be dragons in the Console's keyboard handling stuff that has been under heavy re-architecture of late, so haven't had spare resources to work on this. YET!

We beg your continued patience - we'll get to this ASAP.

/cc @tara-raj @zadjii-msft

**Comment by jskopek** [comment]
Glad to hear this is high on the to-do list!

**Comment by bitcrazed** [comment]
Imagine - we live in the command-line, Console and WSL all day long, every day, and we still haven't gotten around to implementing paste - we feel your pain ;)

**Comment by doggy8088** [comment]
In Command Prompt, it do have an `Shift + Insert` keyboard shortcut to execute paste.   The same keyboard shortcut is used on common linux terminal too.  So I vote for using the same keyboard shortcuts `shift + insert` by default for WSL terminal.

**Comment by zerocool4u2** [comment]
Thank you very much @loxal, here are the scripts that worked for me (ctrl+shift+v and ctrl+v), the program(_AutoHotkey_) also supported the option to create **.exe** files, this allows to skip the installation of the complete software, there are also included in the **zip**, regards!

[bash_paste.zip](https://github.com/Microsoft/WSL/files/1524492/bash_paste.zip)

**Comment by zerocool4u2** [comment]
PD: Here is a bat file to run the exe on start

```bat
@echo off
SET bashPaste=your path to exe here

rem add on start
@reg add "HKEY_LOCAL_MACHINE\SOFTWARE\Wow6432Node\Microsoft\Windows\CurrentVersion\Run" /t REG_SZ /v "bash paste" /d "\"%bashPaste%\"" /f
pause
```
```bat
@echo off

rem remove
@reg delete "HKEY_LOCAL_MACHINE\SOFTWARE\Wow6432Node\Microsoft\Windows\CurrentVersion\Run" /v "bash paste" /f
pause
```

**Comment by avonar** [comment]
Vote up
+1
Please

**Comment by iamakulov** [comment]
Unlike other folks, I’d prefer to have `Ctrl+C`/`Ctrl+V` just like CMD and PowerShell have. I switch between PowerShell and Bash frequently, and having different copy-paste shortcuts would require me to think each time I’m trying to paste something.

`Ctrl+C` does have a special meaning, but we could handle it just like PowerShell does. I.e. if no text is selected, it sends `SIGING`; if some text is selected, it copies it (without sending `SIGING`) and resets the selection.

**Comment by zadjii-msft** [comment]
@iamakulov That's actually how copy works in WSL at the moment. If you have text selected, C-c will copy it.

**Comment by iamakulov** [comment]
@zadjii-msft Right, good point! So only `Ctrl+V` is to be implemented then.

**Comment by avonar** [comment]
i really cant understand why so important thing doesnot fix.
WSL must help linux users and administrators to work in the windows wokspace. @iamakulov said ctrl+c is a good way cause he frequently switch between powershell and bash , so, i freqently switch between linux console and wsl. Changing mind from ctrl+c to shift+insert, makes me crazy, this ticket opened more than one year. But the greatest injustice is that in cmd shift+insert works!

**Comment by pinich** [comment]
If you'r using **Git-Bash** and the 'MINGW64' terminal emulator I found out that you can enable the linux terminal shortcuts like **CTR+Shift+C/V**:

1. Open the option by clicking the left window icon on the left of the window and selecting options
2. Select 'keys' from the tree.
3. Check the **CTR+SHift+letter shortcuts**
4. Click Save or Apply

![2018-01-08 14_14_43-mingw64__c_users_pinich_desktop_tmp_tttttt](https://user-images.githubusercontent.com/11866550/34670494-3492be9a-f47f-11e7-9a14-ae362770178d.png)

**Comment by eromoe** [comment]
@pinich Any idea for `ctrl + u` or `ctrl + w`  cut word to windows clipboard ?

**Comment by bitcrazed** [comment]
@avonar _"I really cant understand why so important thing doesnot fix."_

Appreciate your feedback. We're just as frustrated as you - this issue affects us all day long too!

However, we're a small team with a long prioritized backlog but this item has been lower priority than some other critical work during the last several release cycles. However, we're hopeful that you won't have to wait much longer to see this ask make it above the 'cut' line ;)

Bear with us.

**Comment by DarthSpock** [comment]
My 2 cents on this. Make it configurable per user by user **exactly** the way you can customize keyboard controls on almost all games. No need to battle it out. You choose your own shortcuts in a Option window similar to @pinich has showing and it only pertains to your user account for using cmd/Powershell/bash and customization should include using the shortcuts for all 3 or ability to create separate shortcuts for each. This is not a difficult thing to do and would make many people happy. This has lower priority partly because this isn't a difficult thing to configure - there's be more difficulty working with the other affected parts of MS to get this accomplished but this is something the Console team can create and manage fairly quickly and get released a few Insider upgrades from now. If using TMUX, ConEmu, or some other terminal emulator doesn't function quite the way it did before and is an issue, at that point would it be somewhat in the WSL teams hands along with Console. Those issues if they appear would be quick fixes as well and would place more things on the Release Notes, which in turn puts MS in a good light for showing some progress besides just a couple one liners entered in for a release. It's not very encouraging though it's understood how much work needs to be done and amount of time that might be required for addressing so many issues. A saying comes to mind, "A little bit of effort goes a long way..." MS just needs to show this effort in things like this that are relatively simple.

**Comment by hendrics** [comment]
oh, and i thought i was doing something wrong... i found an not idea workaround: configure bash.exe to be a terminal inside vscode. Then i found ctrl+v to work well and it also gets better colours out of the box. Ofc it is not ideal if you don't use vscode, but might save your some time if you have it installed.

**Comment by cblackuk** [comment]
@hendrics Can you please point me in the right direction as to how to configure this?

**Comment by bitcrazed** [comment]
@DarthSpock Appreciate your analysis of how difficult it is (or is not) to modify the Console and/or WSL codebase. Sorry you think we're not putting much effort into our work, but I can attest from the insanely long hours the entire team has been pouring into these products over the last 3+ years, that you couldn't be further from the truth.

Fact remains that this feature hasn't been prioritized higher because there is an acceptible work-around (right click mouse/trackpad) and because we've had a MOUNTAIN of important, stretegic and tactical work that was of MUCH higher priority for many reasons.

**Comment by iamakulov** [comment]
@bitcrazed Hey! I wanted to say you folks _thank you_ for working on WSL and console and doing your job. Seriously, WSL is an amazing piece of work I love using every day.

It might be easy to lose enthusiasm when your job is criticized so often. I just wanted to remind that many-many more users don’t say anything because they _like_ using your product and don’t have any issues with it. (And they like it because you did your job great.)

❤

**Comment by bitcrazed** [comment]
Many thanks @iamakulov - I and the entire team appreciate your (and everyone else's) support :) We LITERALLY couldn't have made WSL & Console as good as they are today without your support, bugs, and feedback. And we've still got a lot to do yet, so please do keep the feedback coming.

**Comment by ankitkk** [comment]
upvote.

**Comment by n3rd4i** [comment]
There is a way to paste into WSL console using only keyboard, procedure is as follows:
**Requirement: needs to be only ONE WSL window open (at least in combined taskbar buttons)**

1. **WIN + T** -- use this key shortcut until you reach to the WSL window (assuming there is at least one)
2. **SHIFT + F10** -- bring-up context menu for this window
3. **E** + **P** -- Edit and Paste (context menu entries)

**Comment by FredyC** [comment]
@n3rd4i Sadly, the Shift + F10 doesn't work in Powershell window. If there would be another binding for that, this would be a viable approach, although I am not sure you have mentioned Win+T, that's unrelated :)

**Comment by jesseconnell** [comment]
@FredyC It works for me - the `Win+T` cycles focus through the task bar mini-windows (or whatever you call them) - those windows accept `Shift+F10`.  It opens the little shortcut window thingy that `Alt+Space` normally does (or right-clicking the top left corner of the window).

BTW, It would be a usable workaround if `Alt+Space` could optionally be configured to actually get processed normally.  I would guess some action had to be taken to supersede the usual Windows handling of that.  If so, seems like a registry key hack could make it skip that action.

**Comment by FredyC** [comment]
Ah, I see now, that works indeed, but honestly, it might be just faster to grab that mouse and click instead of cycling through all windows with Win+T :)

It seems you don't even need to click on corner icon, right clicking on titlebar works and then you just hit `E + P` and done. Definitely easier :)

**Comment by jesseconnell** [comment]
"might be faster to grab that mouse..."   Oh right!  That's a great workaround - I suggest closing the issue.   :trollface:

**Comment by xdfil** [comment]
I found another solution. I bought a MacBook Pro. I'm not going to wait
around for months/years for Microsoft to make WSL actually useful. Basic
things like copy/paste, really?

MacOS terminal is far better than WSL. File permissions aren't all screwed
up by NTFS either.

I've been making excuses for Windows for most of my life. I'm done.

Use Gnome, or get a Mac.

**Comment by zerocool4u2** [comment]
Hi, three months ago I posted a [zip](https://github.com/Microsoft/WSL/issues/235#issuecomment-348739342) file with scripts for a workaround on ctrl+v and ctrl+shift+v on console, they are really easy to use and honestly they are the more easy workaround I found, thanks again to @sgoranson and @loxal who share how to do this before me.

1. choose an exe file from the zip, ctrl or ctrl+shift
2. test it on bash, you will also get a tray icon to know that it is open, if everything works go to step 5
3. something went wrong or you are using another shell, install autohotkey and edit the files with extension ahk, above are comments about this that might be helpful
4. if you don't want an extra program on your system, you  can compile your code to an exe file on autohotkey and uninstall it
5. everything working? perfect add it to a startup program, to not have to open it every time.
6. here are some bat files to do so

```bat
@echo off
SET bashPaste=your path to exe here

rem add on start
@reg add "HKEY_LOCAL_MACHINE\SOFTWARE\Wow6432Node\Microsoft\Windows\CurrentVersion\Run" /t REG_SZ /v "bash paste" /d "\"%bashPaste%\"" /f
pause
```

7. when this is officially implemented, here is the bat to clean the startup program

```bat
@echo off

rem remove
@reg delete "HKEY_LOCAL_MACHINE\SOFTWARE\Wow6432Node\Microsoft\Windows\CurrentVersion\Run" /v "bash paste" /f
pause
```


I know that is annoying to not have it natively yet, but I'm sure there is a good reason behind it, I share the sentiment of @xdfil if you are a hardcore DevOps or a Pentester, they are a lot of important things that are not implemented yet, but for others who work on back ends like rails for example or many many others fields, this is a game changer, my life has been a lot easier for the past 2 years (06-2016 I believe), when pretty much all I need it was working as expected, been able to not have to jump between OS's is priceless for me, sincerely. Nowadays, I love my console, It could be better?, sure, but every patch I found something new to take advantage and hopefully soon, everything would work and I will be able to hack my neighbor with aircrack.

My current console after some tweaks

![imagen](https://user-images.githubusercontent.com/1785923/37540067-7d903bd8-2934-11e8-95cc-e3bc34319c5e.png)

**Comment by zadjii-msft** [comment]
No need for people to get salty, a little patience goes a long way:

![image](https://user-images.githubusercontent.com/18356694/37686443-b84ab622-2c54-11e8-924b-b885748e28c1.png)

Just checked it in. See y'all in skip-ahead.

**Comment by loxal** [comment]
Nice but when can we expect this feature in the Windows release? Many people got salty because this most basic feature is supported in macOS since a decade whereas Windows took "a bit" longer. Legacy should not be an excuse. I suppose that macOS had to solve similar problems and legacy implementations can be always resolved with explicit opt-in implementations.

**Comment by zadjii-msft** [comment]
The next full release of Windows with this enabled will be whatever comes after Spring Creator's Update (2Fall2Creators Update?).

**Comment by bitcrazed** [comment]
@loxal & @all

We understand the frustration - we feel it all day too because, unsurprisingly, we all live in the Console all day long.

However, simple physics have prevented this feature from bubbling high-enough up our prioritized bug list to receive attention until recently: Console has a small team of 3 who've been working at breakneck speed to add the MANY feature improvements you've all seen over the last 2+ years, adding masses of VT support, 24-bit color, etc. while re-engineering & modernizing a complex 30+ year old code-base.

Legacy may not be an excuse in your mind, but it's a reality that's unavoidable. If we "break" the Console's existing Windows behaviors, we hear from 1000x more users than if we "break" a Linux user's workflow.

> So, understand if there's an issue that appears to be "simple to fix" and it hasn't been, then it's safe to assume that there's a very good reason it's not yet been fixed, frustrating though the delays may be at times.

We greatly appreciate everyone's continued patience and continued feedback & encourage y'all to keep it coming: It helps us find & fix the issues that matter in a sane prioritized order.

**Comment by kavdev** [comment]
> "Legacy should not be an excuse"

@loxal That's actually a pretty damn good excuse in my book. A small team trying to modernize 30+ years of legacy code is a massive undertaking. I wouldn't wish it upon my worst enemy lol

Just think of how many global enterprise users of windows there are. Breaking compatibility in something so foundational as the console framework would be an SLA nightmare.

It's a lot to ask from all of us, but patience is key here. I think we're just all glad to see bash make its way into windows in the first place; look at how far we've come in 2 years!

All that said, @bitcrazed: do you see the Console team getting more resources any time soon or are you guys at optimal stride with 3 people?

**Comment by zadjii-msft** [comment]
> do you see the Console team getting more resources any time soon or are you guys at optimal stride with 3 people?

@kavdev I don't think resource allocation is really something any of us on the Console team are at the pay scale to discuss. I think we'd all love to have more resources, more devs, to be able to churn out more features. Obviously, though - we're all on this team because of how passionate we are about it so of course we want more resources to be able to deliver more features.

It probably doesn't help that we keep rocking it with only 3 devs - what other team has that kind of feature/dev ratio? 😝

**Comment by DarthSpock** [comment]
Idea here: Open-source the console. If Microsoft doesn't want to hire more console devs, then get devs to do it for free! Pretty sure there's been enough enthusiasm gone around to get a bit of help. Of course to avoid legal hassle, they'd probably have to reduce the cost of the OS by X amount since they probably wouldn't want to compensate anyone for their service to a product that they sell that uses the work. Probably still much more complex than that but they've open-sourced Powershell and there's VS Code..plus you've got the ColorTool which is open-source. I would think it's less of a stretch to open-source the console than it is to open-source WSL or even Windows (as great as that would be) but who knows? Maybe it IS a big stretch in MS's mind.

**Comment by bitcrazed** [comment]
@DarthSpock Open-sourcing the console is something we discuss regularly. While it's not on the cards in the near term, it's not entirely out of the realm of possibility. One of the biggest challenges, however, is that the Console is FARRRRR from being in a useful shape for others to rapidly comprehend and contribute to in a meaningful way. Even internally!!

As @zadjii-msft points out above - we'd love to have "some more dev resources" on the team, but at the same time, be they full-time employees or community contributors, additional resources will decrease productivity of an established team - a team who've spent the last 3 years unpicking and understanding, and mastering every corner of a large, old, messy, confusing, but *utterly essential* code-base.

And as @zadjii-msft points out - though it may not be apparent from the outside - the 3 devs we have keep overachieving in every release. Maybe if they struggled and complained more, we might get more resources. 

We'll bear that strategy in mind for the future ;)

**Comment by bitcrazed** [comment]
@kavdev - many thanks :)

We're excited to come into work to improve the Console in tiny/unmeasurable, huge/significant ways each and every day. We're all **passionate** command-line nerds and do this work because we believe Windows users (including us) deserve a kick-a$$ command-line experience. We'll get there - just stick with us while we wrestle this beast into shape :)

**Comment by FredyC** [comment]
It must be a really amazing experience to be on such team, sounds like a well-oiled machine so to speak. Keep up great work!

**Comment by renzok** [comment]
Ohh yeahh! You are awesome!

Although I will wait patiently for this feature to arrive, I have to say 
that I can not wait for it to arrive on my box :)

Until then I like to use the console from within Visual Code as a work 
around for getting ctrl-v/c working.



Am 3/20/2018 um 11:42 PM schrieb Mike Griese:

**Comment by bitcrazed** [comment]
@FredyC  & @renzok  - It is genuinely humbling to work with the awesome WSL and Console engineering teams, and with so many of you in the community. I've learned SO MUCH in the last couple of years and continue to do so daily. 

We're like swans - we appear calm and collected up-top, but are paddling like fury underneath 😁

**Comment by bitcrazed** [comment]
Excited (relieved?) to finally be able to announce that this issue is now fixed (by @zadjii-msft)! [Copy and Paste arrives for Linux/WSL Consoles](https://blogs.msdn.microsoft.com/commandline/2018/04/13/copy-and-paste-arrives-for-linuxwsl-consoles/)

Marked as fixed in insiders build 17643. Will ship in Win10 18H2 release.

Closing since the work is now done.

**Comment by Pixinn** [comment]
That's a great news! Thanks  for all the hard work! 👍

**Comment by shamoons** [comment]
I don't know if this is fixed. I have 1903 and in my Ubuntu bash, CTRL+SHIFT+C/V works, but within WSL for VS Code, it does not.
